### PR TITLE
Hugo migration

### LINKED ISSUE #34: Switch to Hugo?
https://gohugo.io/hosting-and-deployment/

@KurtBruns , would be good if you could list some bullet points of things you like about it. I've personally gotten frustrated with Jekyll more than once, I just chose it because I thought it would be easier for Grant to maintain in the future.

Hugo does seem to have a "jekyll import" function that hopefully works nicely.

### PR DESCRIPTION
**Pull request by vincerubinetti**
Closes #34 

- re-implements the entire site in Hugo, incorp Kurt's feedback
- make folder for each lesson with slug name. rename lesson files index.md. move date to metadata front matter
- remove github action to deploy main branch to github pages. let's switch the live site to Vercel along with the deploy previews.

### CODE REVIEW

**Comment by vincerubinetti** [comment]
Here are my thoughts on Hugo for posterity.

<details><summary>notes</summary>

Hugo seems to have more features, but I wouldn't exactly call it more flexible. In fact I think it's more rigid than Jekyll. Hugo has a lot of specific solutions for problems, where as Jekyll gives more basic building blocks that are less restrictive. Hugo gives an order of magnitude more building blocks that are more specific and restrictive to what you use them for.

Here are some problems I ran into, things that Jekyll can do that Hugo can't.

- partials and shortcodes are distinct. shortcodes can't be used in other templates/partials. passing variables to partials has a bad syntax. so for any component that i want to use both in markdown and in other templates, i have to make it a partial, then also make a shortcode for it that calls the partial.
- no equivalent to the jekyll capture function. no way to capture a bunch of formatted content and pass it to a partial for parsing. i have a partial that takes the main content of the page and processes it; things like wrapping all tables in a div for scrolling. i can pass the "content" variable from a markdown page, but in the lesson template, where there are other things above and below the "content" (like comments, lesson navigation, etc), i can't pass those "above" and "below" things to the same processing function. have to do anything that the processing function does manually for those "above"/"below" sections.
- can't do variable html tag names in a clean way. see the clickable partial for the very ugly way i was able to hack it. the other option would be to duplicate the whole code twice under an if else. if there were more than two types i needed, i'd have to duplicate even more code.
- weird and frequent caching issues with the local hot-reload builds. caused me to think i was doing something wrong often, only to find out i wasn't and just had to restart the server script
- can't put templating in markdown. not super essential, but handy sometimes
- can't put markdown in templates. Jekyll allows this in some circumstances iirc. this helps for example generating ids for headings with markdown. writing headings in html means you have to manually attach section ids
- no auto open CLI option
- no 404 redirect on local. but can still test 404 page by nav'ing right to it
- can't write plugins for it? not sure how relevant this will be
- date formatting is really weird. have to put in a specific date to specify the formatting or it doesn't work. really bizarre design choice. in fact, a lot of small, very obtuse design choices like this that were frustrating

I was able to find work-arounds for all of these though, and they weren't the biggest deal, but they did cause a lot of frustration.

The scope/context concept, passing down the dot, makes a lot of things cleaner. But I found myself making mistakes about what the current dot scope was, as it changes basically every time you go in a nest level. Documentation I found to be comprehensive but not explained well or explained vaguely. The maintainers on the forum seemed to be pretty resistant to fixing the above issues.

Things I liked about Hugo:

- it was indeed very fast. though the site build was only taking a few seconds on Jekyll, and I'm not sure how much that would've increased even when we've added all the markdown content.
- single binary install. should make installing on any machine much easier. also, no gemfile or gem lockfile to sit in the repo.
- the templating, while harder to read, has way more functions than jekyll's does, and it reduced the amount of code a little bit
- the template syntax for including components (shortcodes) is cleaner than jekylls. not a huge deal
- in jekyll, you can't do an html fragment like `</div></section><section><div class="section">` (makes a section break). jekyll didn't like that it started with a closing tag. hugo didn't seem to mind.

Overall I think the switch was worth it enough for the extra features, speed, and slight (?) code reduction. But I probably wouldn't use it again unless there was something Jekyll absolutely couldn't do (sorta unlikely because you can write plugins for it). In fact this has inspired me to look at other solutions besides the two, as I'm not really happy with either of them.

</details>

**Comment by vincerubinetti** [comment]
@3b1b You'll have to change the Vercel configuration to Hugo. Also I deleted the github actions workflow in this PR for deploying the live master site, so we can move to Vercel for that as well.

**Comment by PullJosh** [comment]
> *From notes:* In fact this has inspired me to look at other solutions besides the two [(Jekyll/Hugo)], as I'm not really happy with either of them.

I've been holding back on saying this, because proposing tech changes always has the potential to be a dangerous/time-consuming game. But I *really* love [Next.js](https://nextjs.org/).

* It's the best tool for building fast static sites that use React components...
* ...but it also allows building fully-featured backends so if you wanted to let users sign in someday it would be an easy, incremental upgrade.
* It's built by the Vercel people so everything integrates super nicely.
* Next.js has been has been adopted by a lot of big companies, so I'm confident that it's here to stay as a long-lasting technology.
* And--my favorite feature--it would enable us to use [MDX](https://mdxjs.com/playground), which is fan-*freaking*-tastic. (It's a flavor of markdown that allows you to insert JSX directly in your markdown, instead of liquid tags.)

There's a lot of value in sticking with what we already have. But if you're really unhappy with Jekyll/Hugo (which is frankly understandable), Next is 1000% the correct choice for a project like this one.

(I feel confident that I could recreate 90% of the features of the existing site in Next.js in under 2 days of full-time work. So if this is something that sounds like a good idea, we should consider discussing it more.)

**tl;dr:** Changing technologies is always time consuming and runs the risk of being less of a good idea than expected. But if you're unhappy with Jekyll/Hugo, I will put my full recommendation behind Next.js.

**Comment by vincerubinetti** [comment]
It's not really that I'm unhappy with Hugo or Jekyll for this particular project, it's more that some of the design decisions are just baffling to me. Jekyll is fairly limited but its building blocks are fairly consistent and simple. Hugo works but is more complex and very unintuitive to me.

I had considered tools like next and gatsby, but I think they're just overkill for this project unless the backend and react needs grow by a huge amount. My assumption is that Grant will be maintaining this mostly by himself after a certain point, and will not want to hire someone each time he needs a change. Also assuming he wants to spend more time thinking about math and videos and less time learning more complex technologies.

For now I'd say we just merge this, barring any objections from Grant about the under-the-hood code differences between jekyll and hugo.

**Comment by 3b1b** [comment]
The motive for experimenting with Hugo was mostly a matter of seeing whether it proved to be a tool that was easier for you (i.e. @vincerubinetti) to work with but I'm having trouble inferring whether that's actually the case.  From the perspective of site design, would it be more pleasant to move forward with the familiar but less featureful Jekyll, or the featureful but more "unintuitive" Hugo?

It sounds like from the content creator's side, the main benefits of Hugo come down to an easier installation and shortcodes being a cleaner syntax, which are both a little nice but not game-changers.

As to Next.js, 

Values I have in all of this:
- Clean abstraction between the choice of technology and the content.




> Overall I think the switch was worth it enough for the extra features, speed, and slight (?) code reduction. 

Just to confirm before merging, is there any chance this is sunk cost fallacy?  If it turns out that Jekyll is the simplest



So if I'm reading things right, it sounds like given that the switch to Hugo is written, it has enough marginal benefits to be worth merging in, but in hindsight


Maybe at some point in the future, there will be a project justifying a switch to Next.js, but I agree that its features are probably overkill for this project.  A relevant question, though, is whether it's actually more complex to work with

@PullJosh, supposing that in the future (e.g. 2 years from now) some grander ambition or new project justified a switch from Hugo to Next.js, what's your guess on how much more time it would take to make the switch once we have all of the bloggified videos in place vs. not?  My assumption is that it's not really any more than it would be now, since the content would be abstracted away from the technology,

**Comment by 3b1b** [comment]
The motive for experimenting with Hugo was mostly a matter of seeing whether it proved to be a tool that was easier for you (i.e. @vincerubinetti) to work with but I'm having trouble inferring whether that's actually the case. From the perspective of site design, would it be more pleasant to move forward with the familiar but less featureful Jekyll, or the featureful but more "unintuitive" Hugo?

It sounds like from the content creator's side, the main benefits of Hugo come down to an easier installation, faster builds, and shortcodes being a cleaner syntax, which are all nice to have but not necessarily game-changers.

As to Next.js, I agree its features are overkill for this project, and I also agree there's a risk of endlessly hop scorching from one thing to another.  That said, the one piece of web development technology which I actually _have_ worked with in the past (albeit for a small project) is React, so if you're concerned with protecting the simplicity of my future, don't worry about that.

I'm wary of toggling too much back and forth between technologies, but one question which seems worth asking is if there's any chance next would be nicer to work with on the site design side?  The other relevant question would be if the desire for a little bit of interactivity sprinkled in (e.g. the neural net demos) lends itself better to an environment that is all React anyway, or if any marginal benefits there are outweighed by the complexities of an overkill system for our simple project.

**Comment by KurtBruns** [comment]
@vincerubinetti glad that you made it through the migration OK, hopefully it wasn't too frustrating. It's always easy to forget some of the past gripes with a tool when you make it home, like how weird the date format function is.

Since I'm partially responsible for inciting the migration, I spent some time going over the `hugo` branch and jotted down some notes that may be helpful. I've only included a couple notes that felt **most pertinent** for the sake of time.

>There was a weird thing with slugs. Slug is not a page variable unless you set it in the front matter. Had to duplicate the slug in the front matter so it would set the url correctly. The permalink options let you use "slug", but only if it's set manually. I tried to make each lesson its own section, then have hugo permalink using the deepest most section (resulting in the parent folder name) but it wouldn't work. Jekyll can't do any of this either though.

The [GetPage](https://gohugo.io/functions/getpage/) function returns the context of the page identified by a path in the `content` directory. For example, the following code snippet returns the page at `content/lessons/2019/fourier-series/`:

```
{{ with .Site.GetPage "fourier-series" }}
  ...
{{ end }}
```

Since you are looping through the data in the `topics.yaml` file, I think this should do just what you need instead of identifying the page by its slug. This keeps the `content` directory structure separate from the `public` directory structure, which can be configured using the `permalink` option in the site's config file.

Two additional notes:

1. As an alternative to the `data/topics.yaml`, lessons could be organized into series using a taxonomy.
2. The [GetPage](https://gohugo.io/functions/getpage/) function  can be combined with [Content View Templates](https://gohugo.io/templates/views/) for a nice authoring experience IMO.

In general, managing the dot context and partials is something I've gotten better at with experience.  Like, *sometimes*, you have no choice but to pass variables in addition to the context as discussed in [this issue](https://discourse.gohugo.io/t/how-to-pass-variables-to-a-partial-template/7321) on the forum. I agree that Hugo can feel rigid and obtuse, but the more I've learned and used it, the more I love it.

Obviously I'm biased toward Hugo and this is above my pay grade. I'm happy to share more of my experience if it helps inform the decision process or expedites developing in Hugo land.

**Comment by vincerubinetti** [comment]
I was just about to ping you for a review of the code to make sure I've done things in the most "Hugo" idiomatic way possible. I've tried to. Please review it as much as you have time for. Here are my responses to your notes. Please tell me if I'm missing something. I've spent many hours in the docs and I don't find them to be super clear, but there may be "hidden" methods that I'm not aware of.

> The GetPage function returns the context of the page identified by a path in the content directory. For example, the following code snippet returns the page at content/lessons/2019/fourier-series/:

I will try this out, thanks

> Since you are looping through the data in the topics.yaml file, I think this should do just what you need instead of identifying the page by its slug. 

I don't understand what you mean by this, can you clarify. There needs to be a way to choose what lessons go in which topics (see below) and the slug seems like the more appropriate identifier (short and friendly).

> As an alternative to the data/topics.yaml, lessons could be organized into series using a taxonomy.

I did look into this but it seems like [taxonomies require you to set them in the front matter](https://gohugo.io/content-management/taxonomies#add-taxonomies-to-content) of each lesson. I originally had it this way but Grant wanted to be able to change the order and move the lessons around between different categories from a single file, rather than properties in the front matters.

> The GetPage function can be combined with Content View Templates for a nice authoring experience IMO.

I was excited to use content view templates for the different views of the lesson, but once I looked into them I realized they wouldn't be appropriate because they [can't accept arguments](https://gohugo.io/functions/render/). We have several views of the lessons: full view (just the full lesson page), card view, mini card view, lesson nav view previous, lesson nav view next, tooltip view (coming soon), and possibly more. To do that with a content view, I believe I'd have to duplicate the lesson card code that many times. Instead, I've made it a lesson card partial that accepts size, icon, etc parameters.

Also, note to Grant and Josh, `partials` are basically the equivalent of previously `_includes/template` (under-the-hood stuff you shouldn't have to touch) and `shortcodes` are equivalent to `_includes/components` (stuff you'll use in markdown).

File: start.sh [hunk 1]
```diff
@@ -1 +1 @@
-bundle && bundle exec jekyll serve --trace --open-url --livereload --future
+hugo serve --disableFastRender --buildFuture --buildDrafts --buildExpired
```
Comment (KurtBruns) [inline start.sh:1]: You mentioned some issues with the live reload feature, which I'm guessing is why the `disableFastRender` flag is passed to local server.

If logic lives in a shortcode, that can sometimes be the culprit. See [this issue](https://github.com/gohugoio/hugo/issues/6177) on Hugo's repository. Usually, this isn't a big deal, but it may explain some of your frustrations.

For example, in the file `content/lessons/_index.md` you call a short code to render the lesson gallery. If I had to guess, this may of been where you experienced live reload issues. Moving this logic to a template such as `list.html` should fix this and the `disableFastRender` flag can be removed for faster rendering.

If you were experiencing live reload issues somewhere else, let me know and I'm happy to take a look!

File: layouts/partials/styles.html [hunk 2]
```diff
@@ -0,0 +1,7 @@
+<link rel="stylesheet" href="https://unpkg.com/swiper/swiper-bundle.min.css" />
+
+{{ range (readDir "/assets/css") }}
```
Comment (KurtBruns) [inline layouts/partials/styles.html:3]: If desired, these assets can be bundled + minified for production

File: layouts/partials/clickable.html [hunk 3]
```diff
@@ -0,0 +1,29 @@
+{{- $open_tag := "<button" -}}
+{{- $close_tag := "</button>" -}}
+{{- if .link -}}
```
Comment (KurtBruns) [inline layouts/partials/clickable.html:3]: I would use the [cond](https://gohugo.io/functions/cond/) function here. Maybe something like this.

```
{{ $tag := (cond .link "a" "button") }}
<{{ $tag }} ... >
</{{ $tag }}>
```

You might need to do `(isset . "link")` instead of just `.link` I can't remember off the top of my head.

**Review by KurtBruns — COMMENTED** [review]
I did a general pass over the migration and left comments where they felt appropriate.

One thing related to #40 - I recommend we set up a  [configuration directory](https://gohugo.io/getting-started/configuration/#configuration-directory) structure to manage our `development` and `production` environments. This allows us to have default configuration and then override or have additional configuration per environment.

Then, to build the production environment you would run.

```
hugo --environment production
```

File: content/lessons/_index.md [hunk 4]
```diff
@@ -0,0 +1,3 @@
+### Lessons
+
+{{< lesson-gallery show="topics" >}}
```
[the remark this answers is not in the document]
Comment (KurtBruns) [reply content/lessons/_index.md:3]: This is related to the comment below:

https://github.com/3b1b/3Blue1Brown.com/pull/42#discussion_r638314662

**Comment by vincerubinetti** [comment]
Thank you for your detailed review, going through the notes now. Once I'm done resolving, Grant would you please set up Vercel to test the build before we merge.

File: layouts/partials/lesson-card.html [hunk 5]
```diff
@@ -0,0 +1,34 @@
+<!-- find lesson by slug/id, eg "wallis-product" -->
+{{ $lesson := index (where site.Pages "File.ContentBaseName" "==" .id) 0 }}
```
[the remark this answers is not in the document]
Comment (vincerubinetti) [reply layouts/partials/lesson-card.html:2]: I had to change it to `site.GetPage .id` but this worked and is a lot cleaner.

File: content/lessons/_index.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (vincerubinetti) [reply content/lessons/_index.md:3]: This won't work as the lesson gallery should be passed to the `main` partial for processing and be under the `<main>` element. Also I think it belongs as a shortcode because it's a component that you insert into content. You may want to add explanatory text or images above or below the component (that belong in the markdown). I don't think you can split the main content of the markdown into separate pieces and put them in different places in the template?

File: layouts/partials/lesson-gallery.html [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+      {{ partial "lesson-card" (dict "id" .) }}
+    {{ end }}
+  </div>
+
+  <div class="lesson_gallery_tab" data-tab="topics">
+    {{ range site.Data.topics }}
+      {{ partial "topic-card" (dict "name" .name) }}
+      <div class="lessons">
+        {{ range .lessons }}
+          {{ partial "lesson-card" (dict "id" .) }}
```
[the remark this answers is not in the document]
Comment (vincerubinetti) [reply layouts/partials/lesson-gallery.html:44]: If I understand correctly, here's how this would change things. We'd need:

```
▾ layouts/
  ▾ lessons/
      card.html
      card-mini.html
      card-mini-prev.html
      card-mini-prev.html
      tooltip.html
      single.html
      ...more?
```

Several duplicates, though each file would only be a line or two.

Then this:
```
{{ range site.Data.featured }}
  {{ partial "lesson-card" (dict "id" .) }}
{{ end }}
```

would become something like this?
```
{{ range site.Data.featured }}
  {{ partial "lesson-card" (site.GetPage .) }}
{{ end }}
```

I think this just adds files and doesn't really make anything cleaner.

The scratch pad idea may be workable. But I think it works fine as it is now. lesson-card is a component. If this were react, it'd just be a component that you render with different options in different places.

File: start.sh (hunk 1 above)
Comment (vincerubinetti) [reply start.sh:1]: I can't recall exactly where I ran into these problems, it seemed pretty random. One thing I do remember is that `<title>` in `<head>` wasn't being updated often. I'm certainly going to keep the flag there for my own uses to make sure I never run into that problem.

File: config.yaml [hunk 7]
```diff
@@ -0,0 +1,13 @@
...
+languageCode: en-us
+
+googleAnalytics: "G-MEASUREMENT_ID"
+disqusShortname: yourdiscussshortname
+
+enableInlineShortcodes: true
+enableGitInfo: true
+
+permalinks:
+  lessons: /:section/:slug
```
[the remark this answers is not in the document]
Comment (vincerubinetti) [reply config.yaml:13]: I removed the `slug` property and changed this. Seems to work properly.

File: content/lessons/testbed-lesson.md [hunk 8]
```diff
@@ -1,22 +1,17 @@
...
-tags:
-  - interactive
 ---
 
 Markdown content.
 One sentence per line.
 
-<!-- html comment (will appear in the rendered page's source code) -->
-
-{% comment %}
```
[the remark this answers is not in the document]
Comment (vincerubinetti) [reply content/lessons/testbed-lesson.md:17]: Yeah I took those out because I couldn't think of a case where you guys would want to show a comment in the site's source code. Is there a real use case? Seems like it'd be easy to add later if you ever needed it.

File: content/lessons/testbed-lesson.md [hunk 9]
```diff
@@ -39,12 +34,7 @@ _italic text_
 
 [Link to external page](https://some-website.org/)
 
-[Link to page in root](extras)
-
-[Link to page not in root]({{ "extras/cameos" | relative_url }})
-
-Centered element.
-{:.center}
+[Link to page within site]({{< ref "extras" >}})
```
[the remark this answers is not in the document]
Comment (vincerubinetti) [reply content/lessons/testbed-lesson.md:37]: This is a really cool feature. I put a `render-link` in, so now no fussy rel shortcode.

Funny enough, I ran into a hot-reload issue when changing around `render-link` here. Had to restart the serve.

File: layouts/partials/styles.html (hunk 2 above)
Comment (vincerubinetti) [reply layouts/partials/styles.html:3]: I put in the minify command, but I'd rather keep it unbundled even for production so I can look at individual css files on the live site for debugging purposes, and to save an extra command (or lines). This isn't a strong preference though.

Also the css files are a drop in the bucket compared to all the other image and video files we'll be serving.

File: layouts/partials/clickable.html (hunk 3 above)
Comment (vincerubinetti) [reply layouts/partials/clickable.html:3]: This syntax doesn't work:

image

See this link: https://discourse.gohugo.io/t/using-custom-variables-for-html-tags/15034/7 (Side note, I really despise the moderator saying "Would you rather “readable” or “working” :slight_smile: ?")

And also this was one of my bullet points in my note above.

I did change the code to make use of `cond` though.

**Comment by vincerubinetti** [comment]
Okay finished going through comments. Also incorporated this one:

> One thing related to #40 - I recommend we set up a configuration directory structure to manage our development and production environments. This allows us to have default configuration and then override or have additional configuration per environment.

This makes the root of the repo nice and clean.

File: content/lessons/_index.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (KurtBruns) [reply content/lessons/_index.md:3]: Oh, that makes sense. Like you might want to insert the lesson gallery into the homepage as a shortcode.

>  I don't think you can split the main content of the markdown into separate pieces and put them in different places in the template?

Yeah the main content of the markdown file is kept together. One thing worth mentioning here is that you can structure a page into multiple markdown files using [Page Resources](https://gohugo.io/content-management/page-resources/) when the page is a bundle.

File: layouts/partials/lesson-gallery.html (hunk 6 above)
[the remark this answers is not in the document]
Comment (KurtBruns) [reply layouts/partials/lesson-gallery.html:44]: Definitely fine for now. It would actually look something like:

```
{{ range site.Data.featured }}
  {{ (site.GetPage .).Render "card" }}
{{ end }}
```

I agree that it adds extra files. Although, one long term benefit worth noting is that when you go to update the "lesson-card" partial in some way or another, rather than doing a "find/replace" for all of the places the partial is used, you can just update the content view once (even dramatically) and feel confident that the changes are propagated everywhere the lesson card content view is rendered.

Again, more than fine for now.

File: content/lessons/testbed-lesson.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (KurtBruns) [reply content/lessons/testbed-lesson.md:17]: Easy to add later. I think the use case would be keeping notes on lessons for future selves and other authors.

File: layouts/partials/styles.html (hunk 2 above)
Comment (KurtBruns) [reply layouts/partials/styles.html:3]: Ok, just thought I would mention it.

I think historically reducing HTTP requests was considered good practice. Although, Google mentions in [their performance docs](https://developers.google.com/web/fundamentals/performance/get-started/httprequests-5#a_caveat) that this may not be the case for HTTP/2 requests. No need to try and prematurely optimize.
